contributing.md

```diff
@@ -94,6 +94,118 @@ To check that the package is compatible with server side rendering, build the pr
 yarn test:ssr
 ```
 
+### How to make an entropy source
+
+Entropy source is a function that gets a piece of data about the browser a.k.a. a fingerprint component.
+Entropy sources are located in the [src/sources](src/sources) directory.
+Entropy component must be a simple JS value that can be JSON-encoded.
+
+Entropy source runs in 2 stages: "load" and "get":
+- "Load" runs once when the agent's `load` function is called.
+    It must do as much work of the source as possible to make the "get" phase as fast as possible.
+    It may start background processes that will run indefinitely or until the "get" phase runs.
+- "Get" runs once per each agent's `get` function call.
+    It must be as fast as possible.
+
+An example entropy source:
+
+```js
+// The function below represents the "load" phase
+async function entropySource() {
+  // The "load" phase starts here
+  const preData = await doLongAction()
+
+  // The "load" phase ends when the `entropySource` function returns
+  // The function below represents the "get" phase
+  return async () => {
+    // The "get" phase starts here
+    const finalData = await finalizeData(preData)
+    return finalData
+    // The "get" phase ends then this returned function returns
+  }
+}
+```
+
+Any of the phases can be synchronous:
+
+```js
+function entropySource() {
+  const preData = doLongSynchronousAction()
+
+  return () => {
+    const finalData = finalizeDataSynchronously(preData)
+    return finalData
+  }
+}
+```
+
+The "get" phase can be omitted:
+
+```js
+async function entropySource() {
+  const finalData = await doLongAction()
+
+  // If the source's returned value isn't a function, it's considered as a fingerprint component
+  return finalData // Equivalent to: return () => finalData
+}
+```
+
+In fact, most entropy sources don't require a "get" phase.
+The "get" phase is required if the component can change after completing the load phase.
+
+In order for agent to measure the entropy source execution duration correctly,
+the "load" phase shouldn't run in background (after the source function returns).
+On the other hand, in order not to detain the whole agent, the "load" phase must contain only the necessary actions.
```

Comment:
`detain` is unclear, perhaps you meant "block"?

Reply:
Why is it unclear? Is it a wrong context for the word?

Reply:
detain is when a police detains a criminal, it's like "arrest".